### PR TITLE
Feat: map screen

### PR DESCRIPTION
**Pull request by Tristan-C789**
This PR aims to add the Map screen, a core feature centered around using the GPS sensor.

## Additions 
This PR includes:
- The new screen, properly setup inside the navigation
- Location permission request when needed
- The map, provided by Google Maps
- Custom markers on the maps, corresponding to the challenges that possess a location

## Small fixes
This PR also changes a few logs, adds a new function to retrieve all challenges from the database, or display a warning when no challenge is displayed on the home screen.

### CODE REVIEW

File: app/screens/home/home_screen.tsx [hunk 1]
```diff
@@ -22,7 +23,7 @@ export default function HomeScreen({ user, navigation, firestoreCtrl }: any) {
       const fetchChallenges = async () => {
         try {
           const challengesData = await firestoreCtrl.getChallengesByUserId(uid);
-          console.log("Challenges", challengesData);
+          console.log("Challenges [" + uid + "]", challengesData);
           setChallenges(challengesData);
         } catch (error) {
```
Comment (margdsq) [inline app/screens/home/home_screen.tsx:28]: Maybe you should add a toast or an alert for the user in this case to improve user experience.

File: app/screens/home/home_screen.tsx [hunk 2]
```diff
@@ -47,21 +48,26 @@ export default function HomeScreen({ user, navigation, firestoreCtrl }: any) {
...
-            // Include other props as needed
-          />
-        ))}
+        {challenges.length === 0 ? (
+          <ThemedText>No challenge to display</ThemedText>
+        ) : (
+          challenges.map((challenge, index) => (
+            <Challenge
+              navigation={navigation}
+              firestoreCtrl={firestoreCtrl}
```
Comment (margdsq) [inline app/screens/home/home_screen.tsx:57]: This should include a testID for testing

File: app/screens/map/map_screen.tsx [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+        }}
+        zoomControlEnabled={true}
+        mapType="standard"
+        showsUserLocation={true}
+        showsCompass={true}
+        loadingEnabled={true}
+      >
+        {challengesWithLocation.map((challengeWithLocation: any, index) => (
+          <MapMarker
+            key={index}
```
Comment (margdsq) [inline app/screens/map/map_screen.tsx:121]: To avoid potential problems, it would be better to replace the key with the challenge id (since it is unique). In case of change in order of the elements, it will better handle it.

File: app/screens/map/map_screen.tsx [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+          longitudeDelta: 0.0,
+        }}
+        zoomControlEnabled={true}
+        mapType="standard"
+        showsUserLocation={true}
+        showsCompass={true}
+        loadingEnabled={true}
+      >
+        {challengesWithLocation.map((challengeWithLocation: any, index) => (
+          <MapMarker
```
Comment (margdsq) [inline app/screens/map/map_screen.tsx:120]: Should add test Ids.

File: firebase/FirestoreCtrl.ts [hunk 5]
```diff
@@ -180,4 +180,28 @@ export default class FirestoreCtrl {
       throw error;
     }
   }
+
+  /**
+   * Retrieves all challenges from Firestore. WARNING: This can be slow and expensive.
+   *
+   * @returns A promise that resolves to an array of all challenges.
+   */
```
Comment (margdsq) [inline firebase/FirestoreCtrl.ts:188]: This can easily be overwhelmed if the collection is really large. I guess it's fine for now but maybe in a future PR we could improve this by using another algorithm. For now we can just limit the number of challenges retrieved by using limit in query.

File: firebase/FirestoreCtrl.ts [hunk 6]
```diff
@@ -180,4 +180,28 @@ export default class FirestoreCtrl {
...
+  async getAllChallenges(): Promise<DBChallenge[]> {
+    try {
+      const challengesRef = collection(firestore, "challenges");
+      const querySnapshot = await getDocs(challengesRef);
+      const challenges = querySnapshot.docs.map((doc) => {
+        const data = doc.data();
+        console.log("+ Challenge data retrieved:", data);
```
Comment (margdsq) [inline firebase/FirestoreCtrl.ts:195]: Same here, if the number of challenges is big, it may not be very wise to log something everytime we retrieve a challenge, but for now I think it's okay.

File: __tests__/MapScreen-test.tsx [hunk 7]
```diff
@@ -0,0 +1,35 @@
+import React from "react";
+import { render } from "@testing-library/react-native";
+import MapScreen from "@/app/screens/map/map_screen";
+import * as Location from "expo-location";
+
```
Comment (margdsq) [inline __tests__/MapScreen-test.tsx:5]: There should be tests for edge cases, for example : empty collection, or challenges that don't have all the fields that are valid.

**Review by margdsq — CHANGES_REQUESTED** [review]
I addressed a few points in this review : 
**Tests for Edge Cases:**
Add tests for scenarios like an empty collection or challenges missing certain fields to ensure robustness.
Include a test for when location permissions are denied or getCurrentPositionAsync fails. Improving user experience with a toast or alert in this case would be beneficial.

**Unique Key for Map Rendering:**
Replace the key in challenges.map() with the unique challenge_id to handle changes in order and avoid potential issues.

**Performance Optimization:**
The current approach may not scale well with a large collection. In a future PR, consider optimizing it with a better algorithm or implementing pagination.
For now, limit the number of challenges retrieved.

**Test Accessibility:**
Add testID attributes to key components (e.g., map markers) to facilitate testing.

File: __tests__/MapScreen-test.tsx [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+      { status: "granted" },
+    );
+    (Location.getCurrentPositionAsync as jest.Mock).mockResolvedValue({
+      coords: {
+        latitude: 37.7749,
+        longitude: -122.4194,
+      },
+    });
+  });
+
```
[the remark this answers is not in the document]
Comment (Tristan-C789) [reply __tests__/MapScreen-test.tsx:23]: I'll add more tests to cover edge cases.

File: app/screens/home/home_screen.tsx (hunk 1 above)
Comment (Tristan-C789) [reply app/screens/home/home_screen.tsx:28]: I'm not sure to understand what you mean there, those are raw data to help debug, before displaying in a nicer way that data to the user.

File: app/screens/home/home_screen.tsx (hunk 2 above)
Comment (Tristan-C789) [reply app/screens/home/home_screen.tsx:57]: Despite what GitHub implies, I'm not responsible for the Challenge component. This should be addressed in a different PR.

File: app/screens/map/map_screen.tsx (hunk 3 above)
Comment (Tristan-C789) [reply app/screens/map/map_screen.tsx:121]: I believe it's not an issue, the index is only coming from the TS array, not from Firebase, and is calculated again every time the screen is refreshed. Unless I misunderstood your request, this should not cause any problem.

File: app/screens/map/map_screen.tsx (hunk 4 above)
Comment (Tristan-C789) [reply app/screens/map/map_screen.tsx:120]: Individual test IDs for each marker seems like a layer of complication for not much gain to me. I will try to find a way to test if the markers appear without using test IDs.

File: firebase/FirestoreCtrl.ts (hunk 5 above)
Comment (Tristan-C789) [reply firebase/FirestoreCtrl.ts:188]: Totally agreed with you, I'll see if it's technically positive to introduce a query limit.

File: firebase/FirestoreCtrl.ts (hunk 6 above)
Comment (Tristan-C789) [reply firebase/FirestoreCtrl.ts:195]: This was used for debug purposes and is not needed this precisely anymore. I'll change that to a single log at the end.

File: __tests__/MapScreen-test.tsx (hunk 7 above)
Comment (Tristan-C789) [reply __tests__/MapScreen-test.tsx:5]: I'll add more tests.

**Review by margdsq — APPROVED** [review]
Thank you for your changes.

File: __tests__/MapScreen-test.tsx
Comment (Lil-Noe) [inline __tests__/MapScreen-test.tsx:1]: Thank you for writing a test with mocked values. Eager to see if the mocking pass after the right configuration.

File: app/screens/map/map_screen.tsx
Comment (Lil-Noe) [inline app/screens/map/map_screen.tsx:1]: Commented lines can be deleted for clarity if they do not apport information (ex : imports)
